File: server/utils/queue/defaultPlaylistMetadataManager.js
import { durationFormatter, getDefaultPlaylistMetadataJson, saveDefaultPlaylistMetadataJson, getDefaultPlaylistJson } from "../utils.js";
import BaseQueueManager from "./baseQueueManager.js";

class DefaultPlaylistMetadataManager extends BaseQueueManager {
    constructor() {
        super({
            readFunction: () => {
                const queue = getDefaultPlaylistMetadataJson();
                return queue.map(item => ({
                    ...item,
                    duration: item.duration ? durationFormatter(item.duration) : "00:00"
                }));
            },
            saveFunction: (items) => saveDefaultPlaylistMetadataJson(items),
            validateFunction: (item) => {
                return typeof item === "object" && item.title && item.url;
            },
            formatFunction: (item) => ({
                ...item,
                duration: item.duration ? durationFormatter(item.duration) : "00:00"
            }),
            duplicateCheckKey: "url"
        });
    }

    // Alias methods to match existing API
    addToQueue(item) {
        return this.add(item);
    }

    getFirstFromQueue() {
        return this.getFirst();
    }

    getLastFromQueue() {
        return this.getLast();
    }

    getFilteredData(filters = {}) {
        const allData = super.getAll();
        if (!Object.keys(filters).length) return allData;

        return allData.filter(item => {
            let matches = true;
            
            // Filter by song metadata
            if (filters.urlType && item.urlType !== filters.urlType) matches = false;
            if (filters.playlistId && item.playlistId !== filters.playlistId) matches = false;
            
            // Filter by playlist metadata - requires checking against playlist data
            if (filters.isActive !== undefined || filters.genre) {
                const playlist = this.getPlaylistMetadata(item.playlistId);
                if (playlist) {
                    if (filters.isActive !== undefined && playlist.isActive !== filters.isActive) matches = false;
                    if (filters.genre && playlist.genre !== filters.genre) matches = false;
                } else {
                    matches = false;
                }
            }
            
            return matches;
        });
    }

    getPlaylistMetadata(playlistId) {
        try {
            const playlists = getDefaultPlaylistJson();
            return playlists.find(p => p.playlistId === playlistId);
        } catch (error) {
            console.error('Error getting playlist metadata:', error);
            return null;
        }
    }

    getAll(filters = {}) {
        return this.getFilteredData(filters);
    }

    addManyToQueue(items) {
        return this.addMany(items, false);
    }

    addManyToTop(items) {
        return this.addMany(items, true);
    }
}

export default DefaultPlaylistMetadataManager;

File: app/managers/default_playlist_metadata_manager.py
from app.managers.base_queue_manager import BaseQueueManager
from app.core.utils import (
    duration_formatter, get_default_playlist_metadata_json,
    save_default_playlist_metadata_json, get_default_playlist_json
)

class DefaultPlaylistMetadataManager(BaseQueueManager):
    def __init__(self):
        def read_function():
            queue = get_default_playlist_metadata_json()
            return [
                {
                    **item,
                    'duration': duration_formatter(item.get('duration')) if item.get('duration') else "00:00"
                }
                for item in queue
            ]
        
        def save_function(items):
            return save_default_playlist_metadata_json(items)
        
        def validate_function(item):
            return isinstance(item, dict) and item.get('title') and item.get('url')
        
        def format_function(item):
            return {
                **item,
                'duration': duration_formatter(item.get('duration')) if item.get('duration') else "00:00"
            }
        
        super().__init__({
            'read_function': read_function,
            'save_function': save_function,
            'validate_function': validate_function,
            'format_function': format_function,
            'duplicate_check_key': 'url'
        })
    
    def add_to_queue(self, item):
        return self.add(item)
    
    def get_first_from_queue(self):
        return self.get_first()
    
    def get_last_from_queue(self):
        return self.get_last()
    
    def get_filtered_data(self, filters=None):
        if filters is None:
            filters = {}
        
        all_data = super().get_all()
        if not filters:
            return all_data
        
        filtered = []
        for item in all_data:
            matches = True
            
            if 'urlType' in filters and item.get('urlType') != filters['urlType']:
                matches = False
            if 'playlistId' in filters and item.get('playlistId') != filters['playlistId']:
                matches = False
            
            if 'isActive' in filters or 'genre' in filters:
                playlist = self.get_playlist_metadata(item.get('playlistId'))
                
                if playlist:
                    if 'isActive' in filters and playlist.get('isActive') != filters['isActive']:
                        matches = False
                    if 'genre' in filters and playlist.get('genre') != filters['genre']:
                        matches = False
                else:
                    matches = False
            
            if matches:
                filtered.append(item)
        
        return filtered
    
    def get_playlist_metadata(self, playlist_id):
        try:
            playlists = get_default_playlist_json()
            for p in playlists:
                if p.get('playlistId') == playlist_id:
                    return p
            return None
        except Exception as error:
            return None
    
    def get_all(self, filters=None):
        return self.get_filtered_data(filters)
    
    def add_many_to_queue(self, items):
        return self.add_many(items, False)
    
    def add_many_to_top(self, items):
        return self.add_many(items, True)
